Extend nft samples to pay royalties

NossicaSystems opened the pull request:
## Purpose

Allow an NFT purchase to pay royalties to the minter

## Changes

The nft sample was extended with additional functions as per the standard

## Checklist

- [X] My code follows the style of this project.
- [X] The code compiles without warnings.
- [X] I have performed a self-review of the changes.
- [X] I have documented my code, in particular the intent of the
      hard-to-understand areas.
- [X] (If necessary) I have updated the CHANGELOG.

## CLA acceptance

_Remove if not applicable.

By submitting the contribution I accept the terms and conditions of the
Contributor License Agreement v1.0
- link: https://developers.concordium.com/CLAs/Contributor-License-Agreement-v1.0.pdf

- [X] I accept the above linked CLA.

abizjak commented:
Can you please put on top  of `main` and add the new example to the CI.

Thank you.

NossicaSystems commented:
> Can you please put on top of `main` and add the new example to the CI.
> 
> Thank you.

Done!

DOBEN commented on examples/cis2-multi-royalties/src/lib.rs:
```diff
@@ -0,0 +1,1405 @@
+    /// The identifier of the token.
+    id:      ContractTokenId,
+    /// The sale price.
+    sale_price: u64,
```
In case of arithmetic functions (e.g.`+`, `-`,`/`, ...) work with `TokenAmountU64`, would it make it easier to use `TokenAmountU64` here?

DOBEN commented on examples/cis2-multi-royalties/src/lib.rs:
```diff
@@ -0,0 +1,1405 @@
+    /// The identifier of the payee.
+    id:      Address,
+    /// The royalties to pay.
+    payment: u64,
```
In case of arithmetic functions (e.g.`+`, `-`,`/`, ...) work with `TokenAmountU64`, would it make it easier to use `TokenAmountU64` here?

DOBEN commented on examples/cis2-multi-royalties/src/lib.rs:
```diff
@@ -0,0 +1,1405 @@
+            amount: token_amount,
+            owner: params.owner,
+        }))?;
+
```
Consider adding the `royalty` to the event. The percentage value of how much `royalties` are paid might be interesting to  off-chain tool.

DOBEN commented on examples/cis2-multi-royalties/src/lib.rs:
```diff
@@ -0,0 +1,1405 @@
+    let parameter = TransferParams::from(vec![transfer]);
+
+    // Send back a transfer
+    host.invoke_contract_read_only(
```
Can you add a test to the unit tests below that invokes the `onReceivingCIS2` hook?

NossicaSystems replied:
Isn't this function only called when the receiver is a contract - the receiver is assumed to be an account in this algorithm so the logic for this function has not been changed from the cis2-multi sample?

DOBEN commented on examples/cis2-multi-royalties/src/lib.rs:
```diff
@@ -0,0 +1,1405 @@
+/// Returns the address of the payee and the amount of royalties to pay them
+/// 
+/// It rejects if:
+/// - The contract does not pay royalties
```
```suggestion
/// - This contract does not pay royalties
```

DOBEN commented on examples/cis2-multi-royalties/src/lib.rs:
```diff
@@ -0,0 +1,1405 @@
+/// It rejects if:
+/// - The contract does not pay royalties
+/// - The return result would be 0
+/// - The token is not recognised
```
```suggestion
/// - The token id does not exist in state
```

DOBEN replied:
@NossicaSystems Please address or comment if you want to keep the current text.

NossicaSystems replied:
Updated as suggested

DOBEN commented on examples/cis2-multi-royalties/src/lib.rs:
```diff
@@ -0,0 +1,1405 @@
+    ensure!(host.state().pay_royalty == true, concordium_cis2::Cis2Error::Custom(CustomContractError::ContractDoesNotPayRoyalties));
+
+    let params: CheckRoyaltyParams = ctx.parameter_cursor().get()?;
+    ensure!(params.sale_price > 0, concordium_cis2::Cis2Error::Custom(CustomContractError::NoRoyaltyPayable));
```
```suggestion
    ensure!(params.sale_price > 0, CustomContractError::NoRoyaltyPayable.into());
```

DOBEN commented on examples/cis2-multi-royalties/src/lib.rs:
```diff
@@ -0,0 +1,1405 @@
+
+/// View function to see whether we support paying royalties
+#[receive(contract = "cis2_multi_royalties", name = "supports_interface", return_value = "bool")]
+fn supports_interface<S: HasStateApi>(
```
```suggestion
fn contract_pays_royalties<S: HasStateApi>(
```
Consider using a more descriptive name.

DOBEN replied:
@NossicaSystems Please address or comment the current naming makes sense as well.

NossicaSystems replied:
Updated

DOBEN commented on examples/cis2-multi-royalties/src/lib.rs:
```diff
@@ -0,0 +1,1405 @@
+}
+
+/// View function to see whether we support paying royalties
+#[receive(contract = "cis2_multi_royalties", name = "supports_interface", return_value = "bool")]
```
Consider using a more descriptive name.

DOBEN replied:
@NossicaSystems Please address or comment the current naming makes sense as well.

NossicaSystems replied:
Name changed, apologies.  I thought I had addressed this but must have missed it somehow.

DOBEN commented on examples/cis2-multi-royalties/src/lib.rs:
```diff
@@ -0,0 +1,1405 @@
+        let state = initial_state(&mut builder);
+        let host = TestHost::new(state, builder);
+        let royalty_paid = supports_interface(&ctx, &host);
+        claim_eq!(royalty_paid.unwrap(), false, "Royalty incorrectly initialised");        
```
Instead of `unwrap()` in all the tests, consider using `expect_report()` which will unwrap your function as well but throw a descriptive error in case that is not possible. It has a better way of error reporting than using plain `unwrap()`.

DOBEN replied:
let royalty_paid = supports_interface(&ctx, &host).expect("Invoke should succeed");
        claim_eq!(royalty_paid, false, "Royalty incorrectly initialised");

NossicaSystems replied:
Sorry, I am not sure I understand.  Do you mean I should change royalty_paid.unwrap() to just royalty_paid?  That would not compile unless I also changed the expected result from a bool to an Option.  Maybe I am misreading?

DOBEN commented on examples/cis2-multi-royalties/src/lib.rs:
```diff
@@ -0,0 +1,1405 @@
+    InvokeContractError,
+    /// No royalty payable.
+    NoRoyaltyPayable,
+    /// No royalties not payable in this contract.
```
```suggestion
    /// This contract does not pay royalties.
```

DOBEN commented on examples/cis2-multi-royalties/src/lib.rs:
```diff
@@ -0,0 +1,1405 @@
+    ContractOnly,
+    /// Failed to invoke a contract.
+    InvokeContractError,
+    /// No royalty payable.
```
No royalties can be paid out.

DOBEN commented on examples/cis2-multi-royalties/src/lib.rs:
```diff
@@ -0,0 +1,1405 @@
+/// The parameter for the contract function `init` which sets up the contract.
+#[derive(Serial, Deserial, SchemaType)]
+struct InitParams {
+    /// pay royalties on this contract
```
```suggestion
    /// Specifies if this contract enforces the royalty payout whenever a token is transferred.
```

DOBEN reviewed (changes requested):
I want to highlight that `royalties` are working in some ecosystems rather differently than how this contract works. Just as a note in case, we get questions about it. Both mechanisms have pros/cons. 

E.g. Using Ethereum as an example: Royalties are enforced (or not enforced) by the NFT exchanges, whereas the token contract (aka this contract) transfer's function does not have any logic for the subtraction of royalties (meaning a user can transfer a token to another of its own accounts without paying royalties to the minter). The NFT contract has only a `view` function to query the royalties which is invoked (and then enforced) by NFT exchanges in case the NFT exchange wants to enforce the royalty payout.

DOBEN commented on examples/cis2-multi-royalties/src/lib.rs:
```diff
@@ -0,0 +1,1405 @@
+        }))?;
+
+        // If the receiver is a contract we invoke it.
+        if let Receiver::Contract(address, entrypoint_name) = to {
```
Or add a comment that `royalty_receiver` is an account because we set it to `owner(contractDeployer)`. So no receiveHook function for `royalty_receiver` in this special case needed.

NossicaSystems commented on examples/cis2-multi-royalties/src/lib.rs:
```diff
@@ -0,0 +1,1405 @@
+        if let Receiver::Contract(address, entrypoint_name) = to {
+            let parameter = OnReceivingCis2Params {
+                token_id,
+                amount,
```
I am working through the other comments now (that you for being so thorough, I really appreciate it :-).  However, I think I adjusted the amount to be amount - royalties a few lines about this?  I am happy to refactor to the code if you think moving the calculation here is clearer.

abizjak commented:
@NossicaSystems what is the progress on addressing the comments provided by @DOBEN ? When do you think you will complete that?

NossicaSystems commented:
Hi,

All of them have been done apart from the extra test.  I have been busy on a different high priority CCD project last week and have been on holiday for the passed few days.

I would expect it to be completed early this week.



> On 5 Jun 2023, at 13:08, Aleš Bizjak ***@***.***> wrote:
> 
> ﻿
> @NossicaSystems what is the progress on addressing the comments provided by @DOBEN ? When do you think you will complete that?
> 
> —
> Reply to this email directly, view it on GitHub, or unsubscribe.
> You are receiving this because you were mentioned.

DOBEN commented:
Great. Thanks for the update @NossicaSystems.

abizjak commented:
> Hi, All of them have been done apart from the extra test. I have been busy on a different high priority CCD project last week and have been on holiday for the passed few days. I would expect it to be completed early this week.

Thanks for the update

DOBEN commented on examples/cis2-multi-royalties/src/lib.rs:
```diff
@@ -0,0 +1,1481 @@
+            logger.log(&Cis2Event::Transfer(TransferEvent {
+                token_id,
+                amount: concordium_cis2::TokenAmountU64(royalties.as_ref().unwrap().payment),
+                from: to_address,
```
```suggestion
                from,
```
This should be the `from` address.

NossicaSystems replied:
Good spot

DOBEN commented on examples/cis2-multi-royalties/src/lib.rs:
```diff
@@ -0,0 +1,1481 @@
+    let params: CheckRoyaltyParams = ctx.parameter_cursor().get()?;
+    ensure!(
+        params.sale_price > 0,
+        concordium_cis2::Cis2Error::Custom(CustomContractError::NoRoyaltyPayable)
```
```suggestion
        CustomContractError::NoRoyaltyPayable.into()
```

This appears a couple of times throughout the smart contract. You shorten the syntax by using `into()`.

NossicaSystems replied:
Done

DOBEN commented on examples/cis2-multi-royalties/src/lib.rs:
```diff
@@ -0,0 +1,1481 @@
+    );
+    ensure!(host.state().contains_token(&params.id), ContractError::InvalidTokenId);
+    let token_details =
+        host.state().token_details.get(&params.id).ok_or(ContractError::InvalidTokenId).unwrap();
```
```suggestion
        host.state().token_details.get(&params.id).ok_or(ContractError::InvalidTokenId)?;
```

The `?` operator unwraps the result if `ok` was returned, otherwise it terminates and returns this `InvalidTokenId` error.

NossicaSystems replied:
Done

DOBEN commented on examples/cis2-multi-royalties/src/lib.rs:
```diff
@@ -0,0 +1,1405 @@
+    /// Who minted this token.
+    minter:  Address,
+    /// The percentage royalty.
+    royalty: u64,
```
Since the value is only from 0-100 we could use it's own type or use the lowest type, e.g. `u8`.  No need for such a big number like `u64`.

DOBEN reviewed (commented):
Well done. Your smart contract is looking pretty good. I resolved most comments. we can discuss the few comments that are still open if there is anything unclear. I added a few minor suggestions below.

NossicaSystems commented on examples/cis2-multi-royalties/src/lib.rs:
```diff
@@ -0,0 +1,1481 @@
+        state.transfer(&token_id, amount, &from, &to_address, builder, &royalties)?;
+
+        let mut receivers_amount = amount;
+        if royalties.is_some() {
+            // Log transfer event
+            logger.log(&Cis2Event::Transfer(TransferEvent {
+                token_id,
+                amount: concordium_cis2::TokenAmountU64(royalties.as_ref().unwrap().payment),
+                from: to_address,
+                to: royalties.as_ref().unwrap().royalty_receiver,
+            }))?;
+            receivers_amount -=
+                concordium_cis2::TokenAmountU64(royalties.as_ref().unwrap().payment);
+        }
```
Are you sure these are equivalent?  This change causes a test case to fail.

DOBEN replied:
The test cases need to be updated accordingly.

NossicaSystems commented on examples/cis2-multi-royalties/src/lib.rs:
```diff
@@ -0,0 +1,1481 @@
+
+#[derive(Debug, Serialize, SchemaType)]
+struct CheckRoyaltyResult {
+    /// The identifier of the payee.
```
Done

NossicaSystems commented on examples/cis2-multi-royalties/src/lib.rs:
```diff
@@ -0,0 +1,1481 @@
+                id:         token_id,
+                sale_price: amount.into(),
+            };
+            let royal_result = state.calculate_royalty_payments(royalty_parameters);
+            ensure!(royal_result.is_ok(), ContractError::InvalidTokenId);
+            royalties = Some(royal_result.unwrap());
```
Done

DOBEN commented on examples/cis2-multi-royalties/src/lib.rs:
```diff
@@ -0,0 +1,1468 @@
+            *from_balance -= amount;
+        }
+
+        if royalties.is_some() {
```
You can simplify the syntax to remove the `unwrap()` similar to:

https://github.com/Concordium/concordium-rust-smart-contracts/pull/281#discussion_r1224219364

DOBEN commented on examples/cis2-multi-royalties/src/lib.rs:
```diff
@@ -0,0 +1,1405 @@
+    fn from(c: CustomContractError) -> Self { Cis2Error::Custom(c) }
+}
+
+impl From<NewReceiveNameError> for CustomContractError {
```
@NossicaSystems Please address or comment if I am missing something here.

NossicaSystems replied:
Apologies, I thought I had replied to this previously.  These lines do seem to be used.  If I remove them I get compilation errors.

DOBEN commented on examples/cis2-multi-royalties/src/lib.rs:
```diff
@@ -0,0 +1,1468 @@
+        claim!(result.is_ok(), "Results in rejection");
+
+        // Check the state.
+        let balance0 =
```
Can you check the balance of the three entities `royalty_receiver`, `from_address`, and ` to_address` and use different addresses for all the three, please.

DOBEN reviewed (commented):
Only a few comments left.

DOBEN reviewed (commented):
8 open threads. Please have a look.

NossicaSystems commented on examples/cis2-multi-royalties/src/lib.rs:
```diff
@@ -0,0 +1,1468 @@
+            let temp_royalties = royalties.as_ref().unwrap();
+            let mut royalties_state =
+                self.state.entry(temp_royalties.royalty_receiver).occupied_or(
+                    concordium_cis2::Cis2Error::Custom(CustomContractError::InvalidRoyaltyAddress),
```
This gives an error as royalties needs to be unwrapped.  I can remove the temp variable and use unwrap directly here but the structure is referenced in future lines (315 and 316).  I am happy to change it if requested, I am just want to check I am not missing anything?

NossicaSystems commented on examples/cis2-multi-royalties/src/lib.rs:
```diff
@@ -371,7 +371,9 @@ impl<S: HasStateApi> State<S> {
         let minter = token_details.minter;
-        let royalty_to_pay: u64 = params.sale_price / 100 * token_details.royalty as u64;
+        let royalty_to_pay: u64 =
+            (params.sale_price as f64 * token_details.royalty as f64 / 100.0) as u64;
```
Rats - yes I had forgotten that limitation.  Apologies.

DOBEN commented:
@NossicaSystems Only two unresolved threads left. You can jump to them by clicking on this link as shown:
![UnresolvedThreads](https://github.com/Concordium/concordium-rust-smart-contracts/assets/13039169/5f5ee6f9-12ea-4bf9-a70c-e943242fd6bd)

DOBEN commented on examples/cis2-multi-royalties/src/lib.rs:
```diff
@@ -0,0 +1,1547 @@
+            self.token_details.get(&params.id).ok_or(ContractError::InvalidTokenId)?;
+
+        let minter = token_details.minter;
+        let royalty_to_pay: u64 = params.sale_price * token_details.royalty as u64 / 100;
```
Thinking about it, we should probably prevent overflow here.

```suggestion
        let royalty_to_pay: u64 = params.sale_price.checked_mul(token_details.royalty as u64).ok_or(CustomContractError::Overflow)? / 100;
```
     
Adding an extra error to the `CustomContractError` struct.
``` 
 /// Overflow in royalty calculation.
Overflow,
```


https://doc.rust-lang.org/std/primitive.u64.html#method.checked_mul

NossicaSystems replied:
Good idea - done

DOBEN commented on examples/cis2-multi-royalties/src/lib.rs:
```diff
@@ -0,0 +1,1547 @@
+        claim_eq!(royalty_paid, true, "Royalty incorrectly initialised");
+    }
+
+    // test royalties are initialised properly
+    #[concordium_test]
+    fn test_royalties() {
+        let ctx = TestReceiveContext::empty();
+        let mut builder = TestStateBuilder::new();
+        let state = initial_state(&mut builder);
+        let host = TestHost::new(state, builder);
+        let royalty_paid =
+            contract_pays_royalties(&ctx, &host).expect_report("Invoke should succeed");
+        claim_eq!(royalty_paid, false, "Royalty incorrectly initialised");
+    }
```
This seems to test the same thing as line 1296-1303. Consider removing or clarifying what in addition you are testing here.

NossicaSystems replied:
Have refactored the code to make it clearer that it is testing a positive and negative case and removed duplicated test

DOBEN reviewed (approved):
Please address the last comments and update the PR with the `main` branch (since the `main` branch already contains additional commits). Also, let's wait if @abizjak wants to have a look through before merging.

NossicaSystems commented:
> 

Could you specify which changes you are requesting here, please?

NossicaSystems commented:
> > Could you specify which changes you are requesting here, please?
> 
> I see you have merged `main` into this branch. Can you please address the last three unresolved comments and push the code.

I thought I had already done this, I am not sure what went wrong.  I have repulled and repushed.
